### PR TITLE
Port `freeze` into PackageManager

### PR DESCRIPTION
**Pull request by RulerOfCakes**
After #114, I'm continuing to implement each command into `PackageManager`.

As a relatively simple first change `freeze()` now uses the given state instead of the global variables.

Note that the current implementation of `freeze` still reads from `importlib.metadata.distributions()`, so it isn't completely independent from the global environment. This isn't as trivially separable as the `repodata_*` variables due to certain intrinsic quirks as discussed in the [previous PR](https://github.com/pyodide/micropip/pull/114#discussion_r1648020731), so for now I'm leaving the path/site-packages/install related elements as-is.

### CODE REVIEW

**Comment by RulerOfCakes** [comment]
After the discussion about potentially having separate install directories in the previous PR, I've taken a look at [`tox`](https://github.com/tox-dev/tox) and from a quick overview it looks like their approach is to create an entirely separate python environment with `virtualenv` and avoid juggling with path / site-packages in the same environment. I'm unsure how to potentially incorporate this approach into `micropip` to work with `pyodide` as it doesn't seem to be an easy task, imo it feels like it would be better to leave it as a TODO for now. Any thoughts? @ryanking13

File: micropip/_commands/freeze.py [hunk 1]
```diff
@@ -12,6 +12,12 @@
 
 
 def freeze() -> str:
+    return _freeze(REPODATA_PACKAGES, REPODATA_INFO)
+
+
+def _freeze(
+    repodata_packages: dict[str, dict[str, Any]], repodata_info: dict[str, str]
```
Comment (ryanking13) [inline micropip/_commands/freeze.py:19]: Let's change the parameter names to: `lockfile_packages`, `lockfile_info`.

File: micropip/_commands/freeze.py [hunk 2]
```diff
@@ -23,18 +29,20 @@ def freeze() -> str:
...
     ``lockFileURL`` of :js:func:`~globalThis.loadPyodide`.
     """
-    return json.dumps(freeze_data())
+    return json.dumps(freeze_data(repodata_packages, repodata_info))
 
 
-def freeze_data() -> dict[str, Any]:
-    pyodide_packages = deepcopy(REPODATA_PACKAGES)
+def freeze_data(
+    repodata_packages: dict[str, dict[str, Any]], repodata_info: dict[str, str]
```
Comment (ryanking13) [inline micropip/_commands/freeze.py:36]: Here too.

File: micropip/package_manager.py [hunk 3]
```diff
@@ -4,6 +4,7 @@
 )
 
 from micropip import package_index
+from micropip._commands.freeze import _freeze
```
Comment (ryanking13) [inline micropip/package_manager.py:7]: I don't like `_commands` directory being imported from parent codes. So let's move `_freeze`, `freeze_data` functions to a separate file (e.g. `micropip/freeze.py`) and let both `_commands/freeze.py` and `packagee_manger.py` import from it.

File: tests/test_package_manager.py [hunk 4]
```diff
@@ -27,3 +30,30 @@ def test_set_index_urls():
     finally:
         manager.set_index_urls(default_index_urls)
         assert manager.index_urls == default_index_urls
+
+
+def test_freeze():
+    manager = test_package_manager()
```
Comment (ryanking13) [inline tests/test_package_manager.py:36]:
```suggestion
    manager = PackageManager()
```

calling a test function inside another test function seems a bit weird.

**Review by ryanking13 — COMMENTED** [review]
Thanks @RulerOfCakes!

**Comment by ryanking13** [comment]
> Note that the current implementation of freeze still reads from importlib.metadata.distributions(), so it isn't completely independent from the global environment. This isn't as trivially separable as the repodata_* variables due to certain intrinsic quirks as discussed in the https://github.com/pyodide/micropip/pull/114#discussion_r1648020731, so for now I'm leaving the path/site-packages/install related elements as-is.

Yes, this can be quite complicated, so I think we can work on it separately later, for now you don't need to worry about making the PackageManager work independently.

File: micropip/_commands/freeze.py (hunk 1 above)
Comment (ryanking13) [inline micropip/_commands/freeze.py:19]: (No action required for this PR)

For better typing and test coverage, we can think of using pyodide-lock package (which generates these repodata lock file) #88.

```python
from typing import TYPE_CHECKING

if TYPE_CHECKING:
  from pyodide_lock import PackageSpec, InfoSpec
else:
  PackageSpec = ...
  InfoSpec = ...

_freeze(packages: PackageSpec, info: InfoSpec)
```

File: tests/test_package_manager.py (hunk 4 above)
Comment (RulerOfCakes) [reply tests/test_package_manager.py:36]: True, I'll change `test_package_manager` to a non-test function.

File: tests/test_package_manager.py [hunk 5]
```diff
@@ -27,3 +30,30 @@ def test_set_index_urls():
     finally:
         manager.set_index_urls(default_index_urls)
         assert manager.index_urls == default_index_urls
+
+
+def test_freeze():
+    manager = get_test_package_manager()
+
+    test_repodata_info = {
```
Comment (ryanking13) [inline tests/test_package_manager.py:38]: Later, I think we can improve this by using an actual lock file structure.

**Review by ryanking13 — APPROVED** [review]
Thanks @RulerOfCakes!

**Comment by hoodmane** [comment]
Thanks @RulerOfCakes!

**Pull request merged**